cirq-google/cirq_google/workflow/quantum_executable.py

```diff
@@ -0,0 +1,157 @@
+import abc
+import dataclasses
+from dataclasses import dataclass
+from typing import Union, Tuple, Optional, Sequence, cast
+
+import cirq
+from cirq import NamedTopology, _compat
+from cirq.protocols import dataclass_json_dict
+from cirq.study.resolver import _is_param_resolver_or_similar_type
+
+
+class ExecutableSpec(metaclass=abc.ABCMeta):
+    """Specification metadata about an executable.
+
+    Subclasses should add problem-specific fields.
+    """
+
+    executable_family: str = NotImplemented
+    """A unique name to group executables."""
+
+
+@dataclass(frozen=True)
+class BitstringsMeasurement:
+    """Use in-circuit MeasurementGate to collect many repetitions of strings of bits.
+
+    This is the lowest-level measurement type allowed in `QuantumExecutable` and behaves
+    identically to the `cirq.Sampler.run` function. The executable's circuit must contain
+    explicit measurement gates.
+
+    Args:
+        n_repeitions: The number of repetitions to execute the circuit.
+    """
+
+    n_repetitions: int
+
+    def _json_dict_(self):
+        return dataclass_json_dict(self, namespace='cirq.google')
+
+    def __repr__(self):
+        return _compat.dataclass_repr(self, namespace='cirq_google')
+
+
+TParamPair = Tuple[cirq.TParamKey, cirq.TParamVal]
+
+
+@dataclass(frozen=True)
+class QuantumExecutable:
+    """An executable quantum program.
+
+    This serves a similar purpose to `cirq.Circuit` with some key differences. First, a quantum
+    executable contains all the relevant context for execution including parameters as well as
+    the desired number of repetitions. Second, this object is immutable. Finally, there are
+    optional fields enabling a higher level of abstraction for certain aspects of the executable.
+
+    Attributes:
+        circuit: A circuit describing the quantum operations to execute.
+        measurement: A description of the type of measurement. A valid option is to use
+            MeasurementGate in your circuit and specify
+            `measurement=BitstringsMeasurement(n_repetitions)`. Future changes may permit
+            different types of measurement.
```

Comment:
Nit: Could we tighten this up to something like:

"A cirq.google.BitStringMeasurement describing the measurement properties/process" or something like that ? If we want to add more types later we should do so when the API gets updated and not leave any ambiguity as to if/when new things may or may not be changing in the API.

Reply:
Sure. I wanted to head off any questions of the form "why are we wrapping one int in its own class" questions from any reviewers, but now you know :)